src/main/java/com/google/devtools/build/lib/bazel/repository/PatchUtil.java

```diff
@@ -0,0 +1,457 @@
+    int oldLineCount = 0;
+    int newLineCount = 0;
+    Result result;
+
+    for (int i = 0; i < patch.size(); i++) {
+      String line = patch.get(i);
+      LineType type;
+      switch (type = getLineType(line, isReadingChunk)) {
+        case OLD_FILE:
+          patchContent.add(line);
+          oldFile = getFilePath(line, strip, outputDirectory);
+          break;
+        case NEW_FILE:
+          patchContent.add(line);
+          newFile = getFilePath(line, strip, outputDirectory);
+          break;
+        case CHUNK_HEAD:
+          int pos = line.indexOf("@@", 2);
+          String headerStr = line.substring(0, pos + 2);
+          patchContent.add(headerStr);
+          header = new ChunkHeader(headerStr);
+          oldLineCount = 0;
+          newLineCount = 0;
+          isReadingChunk = true;
+          break;
+        case CHUNK_ADD:
+          newLineCount++;
+          patchContent.add(line);
+          result = header.check(oldLineCount, newLineCount);
+          if (result == Result.COMPLETE) {
+            isReadingChunk = false;
+          } else if (result == Result.ERROR) {
+            throw new PatchFailedException(
+                "Wrong chunk detected near line " + (i + 1) + ": " + line +
+                    ", does not expect an adding line here.");
+          }
+          break;
+        case CHUNK_DEL:
+          oldLineCount++;
+          patchContent.add(line);
+          result = header.check(oldLineCount, newLineCount);
+          if (result == Result.COMPLETE) {
+            isReadingChunk = false;
+          } else if (result == Result.ERROR) {
+            throw new PatchFailedException(
+                "Wrong chunk detected near line " + (i + 1) + ": " + line +
+                    ", does not expect a deleting line here.");
+          }
+          break;
+        case CHUNK_EQL:
+          oldLineCount++;
+          newLineCount++;
+          patchContent.add(line);
+          result = header.check(oldLineCount, newLineCount);
+          if (result == Result.COMPLETE) {
+            isReadingChunk = false;
+          } else if (result == Result.ERROR) {
+            throw new PatchFailedException(
+                "Wrong chunk detected near line " + (i + 1) + ": " + line +
+                    ", does not expect an equal line here.");
+          }
+          break;
+        case RENAME_FROM:
+          hasRenameFrom = true;
+          if (oldFile == null) {
+            // len("rename from ") == 12
+            oldFile = outputDirectory.getRelative(line.substring(12));
+          }
+          break;
+        case RENAME_TO:
+          hasRenameTo = true;
+          if (newFile == null) {
+            // len("rename to ") == 10
+            newFile = outputDirectory.getRelative(line.substring(10));
+          }
+          break;
+        case GIT_LINE:
+          break;
+        case GIT_HEADER:
+        case UNKNOWN:
+          boolean isRenaming = isGitDiff && hasRenameFrom && hasRenameTo;
```

Comment:
1) Is that not an error that "isRenaming" is only for Git case?
2) It should be a comment here that UNKNOWN is just for a "not specific" line, context line.
Or rename UNKNOWN into CONTEXT... Otherwise it is intimidating to see so much complicated code for processing UNKNOWN case :)
3) Should we have an "else" clause here? It looks like so many conditions from previous lines should match for chunk to be applied, so if some patch text is corrupted, we will silently not apply a chunk... 
Maybe just see if the chunk was not applied and report that.
I know that we kind of checked other conditions already as we read the lines, but the code will be more resilient this way.

Reply:
renaming is a git specific format, so it must come with "diff --git", I hope I had made the UNKNOWN part more clear.

Reply:
Great! I really like it.
A couple more questions:
- what of by some reason the patch format is incorrect and "renameFrom" part is missing? In that case, isRenaming will be false, and we do not rename the file and silently skip all the information about an intention to rename that file in a patch.
We should instead throw an error about patch format.

Also, if we encounter some "unknown" patch lines when we saw no description header before - should we give an error as well?

This is what I meant when I wrote the comment here suggesting to have an "else" section here.

Reply:
So, if we don't see all those three lines (`diff --git `, `rename from ` and `rename to `), we cannot be sure this patch intends to renaming a file, so we should just apply it as a normal patch. I tested with the native patch tool, if I remove any of the lines, it will not do the renaming but will still apply the patch.

And for the second question, a patch file can actually contain any random messages at the beginning or end of the file, or even between two patches. You can see an example by generating a patch file through this command `git format-patch HEAD~2 --stdout > my.patch`. Those additional lines should be ignored silently, this is also the patch command line's behavior. 

The current implementation tries to recognized patch chunks (`@@ +<num>,<num> -<num>,<num>@@`) and "renaming format". It does give error when it saw those formats but didn't get enough information. For other completely irrelevant lines, I think we should just ignore them.

Reply:
Thank you for the explanation, it makes much sense to me now.

What I would do in this case is have an build event with the skipped patch fragments information.
(In the case of the operating system patch, there is  --verbose  Output extra information about the work being done. option, so have something similar.)

The reason being making automatic control over the patch process possible.
If you want, we could put it as a separate GitHub issue.